addons/cert.py

```diff
@@ -234,6 +234,24 @@ def msc30(data):
         if simpleMatch(token, "rand ( )") and isStandardFunction(token):
             reportError(token, 'style', 'Do not use the rand() function for generating pseudorandom numbers', 'MSC30-c')
 
+# STR03-C
+# Do not inadvertently truncate a string
+def str03(data):
+    for token in data.tokenlist:
+        if simpleMatch(token, 'strncpy ('): # now search if the 3rd parameter is a sizeof()
+            paramToken = token.astParent
+            if paramToken is None:
+                continue
+            if not paramToken.astOperand2 is None:
+                nextOp2 = paramToken.astOperand2
+                if nextOp2.astOperand2 is None:
+                    continue
+                lengthOp2 = nextOp2.astOperand2
+                if lengthOp2 is None:
+                    continue
+                if simpleMatch(lengthOp2.astOperand1, 'sizeof ('):  
+                    reportError(token, 'style', 'Do not inadvertently truncate a string', 'STR03-C')
```

Comment:
hmm.. isn't it typical to use a sizeof in the 3rd argument when using strncpy. I fear that this will warn about typical strncpy usage.

Reply:
We do allow more noise from addons. But I am uncertain about how much we can allow.

Reply:
Well the check fullfills rule CERT-STR03-C as defined here: https://wiki.sei.cmu.edu/confluence/display/c/STR03-C.+Do+not+inadvertently+truncate+a+string
The reason for this check is that if you use strncpy with a sizeof() as length it could happen that there isn't a null termination in the first n characters you want to copy. This would lead to a non null-terminated string but strncpy/strcpy normaly indicates that the string is automatically null-terminated ;)

Reply:
I have a bad feeling about it but I guess we should follow the cert guidelines.